lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py

```diff
@@ -0,0 +1,296 @@
+__metaclass__ = type
+
+# TODO
+# - make unit tests
+# - test what happens when the first host is down, goes to second
+# - test what happens if all hosts are down
+
+ANSIBLE_METADATA = {
+    'metadata_version': '1.1',
+    'status': ['preview'],
+    'supported_by': 'community'
+}
+
+
+DOCUMENTATION = '''
+---
+module: vmware_datastore_upload
+short_description: Upload a local file to a VMware datastore.
+description:
+    - Upload a local file to a VMware datastore with the option to
+      skip the upload if the file already exists and is the same.
+author:
+    - "James E. King III (@jeking3) <jking@apache.org>"
+version_added: 2.8
+notes:
+    - Tested on vSphere 6.7
+requirements:
+    - python >= 2.6
+    - PyVmomi
+    - requests >= 2.18.4
+options:
+    datacenter:
+      description:
+      - Name of the datacenter containing the datastore.
+      - If not specified, and more than one datastore with the
+        same name exists, the action will fail.
+      required: no
+    datastore:
+      description:
+      - Name of datastore that will store the file.
+      required: yes
+    source:
+      description:
+      - The full path to the local file to push to vCenter.
+      required: yes
+    destination:
+      description:
+      - The full path to the destination file in the datastore.
+      required: yes
+    debounce:
+      description:
+      - Uploading large files unnecessarily can be avoided
+        with the C(debounce) option.
+      - C(hash) will hash C(source) (using SHA256) and check
+        for a file named C(<destination>.sha256) - if that file
+        does not exist, or the hash in that file differs from
+        the local hash, the file is uploaded and the hash is
+        placed into C(<destination>.sha256) to debounce the next
+        upload.  This option requires no write access locally.
+      - C(hashcache) is the same as C(hash) except that it requires
+        local write access to C(<source>.sha256) so the hash can
+        be cached locally.  If C(source) is modified then the hash
+        will be recalculated on the next upload.  This option
+        prevents the local server from re-hashing large data files.
+      - If C(none), the file will always be uploaded and no
+        C(.sha256) files will be used.  This is appropriate
+        for smaller files.
+      default: none
+      choices: [ none, hash, hashcache ]
+extends_documentation_fragment: vmware.documentation
+'''
+
+EXAMPLES = '''
+- name: Upload, overwriting any existing file at the destination.
+  vmware_datastore_upload:
+    hostname: '{{ vcenter_hostname }}'
+    username: '{{ vcenter_username }}'
+    password: '{{ vcenter_password }}'
+    datastore: '{{ datastore_name }}'
+    source: '{{ source_file_name }}'
+    destination: '{{ destination_file_name }}'
+  delegate_to: localhost
+'''
+
+RETURN = '''
+results:
+    description: result of the operation
+    returned: always
+    type: dict
+    sample: { "changed": false }
```

Comment:
Would it be possible to add a verbose message? 
E.g. `"msg": File 'abc' already uploaded` or `"msg": File 'abc' uploaded`

Reply:
+1 for this.

Reply:
Yes, can do.